File: backend/models/pitcher.py
from pydantic import BaseModel


class PitcherTraditionalStats(BaseModel):
    ERA: float
    IP: float
    H: int
    R: int
    ER: int
    HR: int
    BB: int
    K: int
    W: int
    L: int
    SV: int


class PitcherStatsAgainst(BaseModel):
    BA: float
    OBP: float
    SLG: float
    OPS: float
    BAbip: float
    WHIP: float
    FIP: float
    ERAPlus: int


class PitcherExpectedRate(BaseModel):
    xBA: float
    xSLG: float
    xwOBA: float
    xFIP: float
    H9: float
    HR9: float
    BB9: float
    K9: float
    kBB: float
    HRPct: float
    KPct: float
    BBPct: float


class PitcherStatcastStats(BaseModel):
    exitVelo: float
    launchAngle: float
    hardHitPct: float
    barrelPct: float
    sweetSpotPct: float


class PitcherPercentiles(BaseModel):
    xBA: int
    xSLG: int
    xwOBA: int
    barrelPct: int
    kPct: int
    bbPct: int
    whiffPct: int
    chasePct: int
    exitVelo: int
    hardHitPct: int


class PitchUsage(BaseModel):
    FB: float | None
    SL: float | None
    CB: float | None
    CH: float | None
    SI: float | None
    FC: float | None
    ST: float | None
    SLV: float | None
    KN: float | None
    FS: float | None


class Pitcher(BaseModel):
    id: int
    player_id: int
    name: str
    team: str
    number: int
    position: str
    throw: str
    bat: str
    traditional: PitcherTraditionalStats
    statsAgainst: PitcherStatsAgainst
    expectedRate: PitcherExpectedRate
    statcastAdv: PitcherStatcastStats
    percentiles: PitcherPercentiles
    usage: PitchUsage

File: frontend/src/types/pitcher.ts
export interface PitcherTraditionalStats {
  ERA: number;
  IP: number;
  H: number;
  R: number;
  ER: number;
  HR: number;
  BB: number;
  K: number;
  W: number;
  L: number;
  SV: number;
}

export interface PitcherStatsAgainst {
  BA: number;
  OBP: number;
  SLG: number;
  OPS: number;
  BAbip: number;
  WHIP: number;
  FIP: number;
  ERAPlus: number;
}

export interface PitcherExpectedRate {
  xBA: number;
  xSLG: number;
  xwOBA: number;
  xFIP: number;
  H9: number;
  HR9: number;
  BB9: number;
  kBB: number;
  K9: number;
  HRPct: number;
  KPct: number;
  BBPct: number;
}

export interface PitcherStatcastAdvanced {
  exitVelo: number;
  launchAngle: number;
  hardHitPct: number;
  barrelPct: number;
  sweetSpotPct: number;
}

export interface PitcherPercentiles {
  xBA: number;
  xSLG: number;
  xwOBA: number;
  barrelPct: number;
  kPct: number;
  bbPct: number;
  whiffPct: number;
  chasePct: number;
  exitVelo: number;
  hardHitPct: number;
}

export interface PitchUsage {
  FB: number | null;
  SL: number | null;
  CB: number | null;
  CH: number | null;
  SI: number | null;
  FC: number | null;
  ST: number | null;
  SLV: number | null;
  KN: number | null;
  FS: number | null;
}

export interface Pitcher {
  id: number;
  player_id: number;
  name: string;
  team: string;
  number: number;
  position: string;
  throw: string;
  bat: string;
  traditional: PitcherTraditionalStats;
  statsAgainst: PitcherStatsAgainst;
  expectedRate: PitcherExpectedRate;
  statcastAdv: PitcherStatcastAdvanced;
  percentiles: PitcherPercentiles;
  usage: PitchUsage;
}
